code/modules/power/turbine.dm

```diff
@@ -0,0 +1,431 @@
+
+	if(default_deconstruction_crowbar(user, attacking_item))
+		return TRUE
+
+	if(default_deconstruction_screwdriver(user, attacking_item))
+		return TRUE
+
+	return ..()
+
+/obj/machinery/power/compressor/proc/trigger_overheat()
+	starter = FALSE
+	last_overheat = world.time
+	overheat -= 50
+	radio.autosay(OVERHEAT_MESSAGE, name, "Engineering")
+	playsound(src, 'sound/machines/buzz-two.ogg', 100, FALSE, 40, 30, falloff_distance = 10)
+
+/obj/machinery/power/compressor/proc/time_until_overheat_done()
+	return max(last_overheat + OVERHEAT_TIME - world.time, 0)
+
+/obj/machinery/power/compressor/process()
+	if(!turbine)
+		stat |= BROKEN
+	if(!starter)
+		return
+	ClearOverlays()
+	if(stat & BROKEN || panel_open)
+		return
+
+	if(rpm_threshold == OVERDRIVE)
+		overheat += 2
+		if(overheat >= OVERHEAT_THRESHOLD)
+			trigger_overheat()
+		else if(overheat > 0)
+			overheat -= 2
+	rpm = 0.9 * rpm + 0.1 * rpmtarget
+	var/datum/gas_mixture/environment = inturf.return_air()
+	var/transfer_moles = environment.total_moles/10
+	var/datum/gas_mixture/removed = environment.remove(transfer_moles)
+	gas_contained.merge(removed)
+
+	// RPM function to include compression friction - be advised that too low/high of a compfriction value can make things screwy
+	rpm = max(0, rpm - (rpm*rpm)/(COMPFRICTION*efficiency))
+
+	if(!(stat & NOPOWER))
+		draw_power(2800)
+		if(rpm < 1000)
+			rpmtarget = 1000
+	else
+		if(rpm < 1000)
+			rpmtarget = 0
+
+	var/new_rpm_threshold
+	switch(rpm)
+		if(50001 to INFINITY)
+			new_rpm_threshold = OVERDRIVE
+		if(10001 to 50000)
+			new_rpm_threshold = VERY_FAST
+		if(2001 to 10000)
+			new_rpm_threshold = FAST
+		if(501 to 2000)
+			new_rpm_threshold = SLOW
+		else
+			new_rpm_threshold = NONE
+
+	if(rpm_threshold != new_rpm_threshold)
+		rpm_threshold = new_rpm_threshold
+		AddOverlays(overlay_image(icon, "comp_o[rpm_threshold]", FLY_LAYER))
+		AddOverlays(emissive_appearance(icon, "comp_o[rpm_threshold]", FLY_LAYER))
+		update_icon()
+
+// These are crucial to working of a turbine - the stats modify the power output.
+// TURBPOWER modifies how much raw energy can you get from rpms,
+// TURBCURVESHAPE modifies the shape of the curve - the lower the value the less straight the curve is.
+
+#define TURBPOWER 500000
+#define TURBCURVESHAPE 0.5
+#define POWER_CURVE_MOD 1.7 // Used to form the turbine power generation curve
+
+/obj/machinery/power/turbine/Initialize(mapload)
+	. = ..()
+// The outlet is pointed at the direction of the turbine component
+	outturf = get_step(src, dir)
+	locate_machinery()
+
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/power/turbine/RefreshParts()
+	var/P = 0
+	for(var/obj/item/stock_parts/capacitor/C in component_parts)
+		P += C.rating
+	productivity = P / 6
+
+/obj/machinery/power/turbine/proc/locate_machinery()
+	if(compressor)
+		return
+	compressor = locate() in get_step(src, get_dir(outturf, src))
+	if(compressor)
+		compressor.locate_machinery()
+
+/obj/machinery/power/turbine/process()
```

Comment:
process needs the parameter `ticks_per_seconds` as parent and to multiply everything related to time with it